crates/driver/src/domain/competition/solution/trade.rs

```diff
@@ -25,61 +25,119 @@ pub struct Fulfillment {
-                .ok_or(Error::ClearingPriceMissing)?
-                .to_owned();
-            let buy_price = clearing_prices
-                .0
-                .get(&self.order.buy.token)
-                .ok_or(Error::ClearingPriceMissing)?
-                .to_owned();
-            match self.order.side {
-                order::Side::Buy => (buy_price, sell_price),
-                order::Side::Sell => (sell_price, buy_price),
-            }
-        };
-        let output = eth::Asset {
-            amount: match self.order.kind {
-                order::Kind::Market => input
-                    .amount
-                    .checked_mul(input_price)
-                    .ok_or(Error::Overflow)?
-                    .checked_ceil_div(&output_price)
-                    .ok_or(Error::Overflow)?,
-                order::Kind::Limit { .. } => todo!(),
-                order::Kind::Liquidity => todo!(),
+        // Values needed to calculate the executed amounts.
+        let ExecutionParams {
+            side,
+            kind,
+            sell,
+            buy,
+            executed,
+        } = match self {
+            Trade::Fulfillment(trade) => ExecutionParams {
+                side: trade.order.side,
+                kind: trade.order.kind,
+                sell: trade.order.sell,
+                buy: trade.order.buy,
+                executed: trade.executed,
             },
-            token: match self.order.side {
-                order::Side::Buy => self.order.sell.token,
-                order::Side::Sell => self.order.buy.token,
+            Trade::Jit(trade) => ExecutionParams {
+                side: trade.order.side,
+                // For the purposes of calculating the executed amounts, a JIT order behaves the
+                // same as a regular market order.
+                kind: order::Kind::Market,
+                sell: trade.order.sell,
+                buy: trade.order.buy,
+                executed: trade.executed,
             },
         };
-        Ok(Execution { input, output })
+
+        // Clearing prices.
+        let sell_price = clearing_prices
+            .0
+            .get(&sell.token)
+            .ok_or(Error::ClearingPriceMissing)?
+            .to_owned();
+        let buy_price = clearing_prices
+            .0
+            .get(&buy.token)
+            .ok_or(Error::ClearingPriceMissing)?
+            .to_owned();
+
+        // Calculate the executed amounts.
+        Ok(match side {
+            order::Side::Buy => Execution {
+                buy: eth::Asset {
+                    amount: executed.into(),
+                    token: buy.token,
+                },
+                sell: eth::Asset {
+                    amount: match kind {
+                        // TODO Is this really something that should return an error?
+                        // Just use BigInt for the calculation instead and the FINAL result needs
+                        // to be able to fit into a U256
```

Comment:
Erroring on intermittent overflows mimics the behavior of the smart contract. **Technically** we don't have to do that here but I think it's still useful as the solver would propose a solution that would not work in the settlement contract. We would still detect overflows of the actual settlement during the simulation but I don't see why we shouldn't do it here as well.